### PR TITLE
Add framework for adding new attn bias variants

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
[:copilot](
### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 194e541</samp>

This pull request improves the attention functionality in PyTorch by adding new tests, benchmarks, and documentation for the `torch.nn.utils.attention` module. It also fixes a minor bug in `sdp_utils.cpp` and clarifies the `is_causal` argument of the `scaled_dot_product_attention` function.)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113103

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Comment by drisspg** [comment]
@mlazos 
``` Python
import torch

from torch.nn.attention import TensorBias

@torch.compile(fullgraph=True)
def main():
    q = torch.rand(1, 1, 16, 16, dtype=torch.float16, device="cuda")
    k = torch.rand(1, 1, 16, 16, dtype=torch.float16, device="cuda")
    v = torch.rand(1, 1, 16, 16, dtype=torch.float16, device="cuda")

    bias = TensorBias(torch.rand(1, 1, 16, 16, dtype=torch.float16, device="cuda"))

    out = torch.nn.functional.scaled_dot_product_attention(q, k, v, bias)
if __name__ == "__main__":
    main()
    
    ```
    
    currently produces
    ```Shell
        unimplemented(
  File "/home/drisspg/meta/pytorch/torch/_dynamo/exc.py", line 192, in unimplemented
    raise Unsupported(msg)
torch._dynamo.exc.Unsupported: torch.* op returned non-Tensor TensorBias call_function <class 'torch.nn.attention.TensorBias'>

from user code:
   File "/home/drisspg/meta/scripts/sdpa/subclass.py", line 11, in main
    bias = TensorBias(torch.rand(1, 1, 16, 16, dtype=torch.float16, device="cuda")
    ```

**Review by mikaylagawarecki — COMMENTED** [review]
From triage meeting: this can live in `torch.nn.utils.attention`

File: torch/nn/utils/attention.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+        args = pytree.tree_map_only(
+            TensorBias, lambda x: materialize_if_needed(x), args
+        )
+        kwargs = pytree.tree_map_only(
+            TensorBias, lambda x: materialize_if_needed(x), kwargs
+        )
+        return func(*args, **kwargs)
+
+
+class LambdaBias(AttnBias):
```
Comment (drisspg) [inline torch/nn/utils/attention.py:105]: lets remove this until I actually implement it in follow up PR

File: torch/nn/utils/attention.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+        value: torch.Tensor,
+        attn_mask,
+        dropout_p: float,
+        is_causal: bool,
+        scale: Optional[float],
+    ) -> torch.Tensor:
+        raise NotImplementedError("This is an abstract base class")
+
+
+class TensorBias(AttnBias):
```
Comment (drisspg) [inline torch/nn/utils/attention.py:61]: originally I was going to have this because I wanted to push all people to using this for creating attention biases. Now that we are using __torch__function__ I don't know if the added overhead makes sense. As well I have concerns around compile + export + every other subsystem

cc @albanD

File: torch/nn/utils/attention.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+class AttnBias(ABC):
+    """Abstract base class for attention biases"""
+
+    @abstractmethod
+    def materialize(self, device: Optional[torch.device] = None) -> torch.Tensor:
+        raise NotImplementedError("This is an abstract base class")
+
+    @abstractmethod
+    def needs_materialization(self) -> bool:
+        raise NotImplementedError("This is an abstract base class")
```
Comment (jbschlosser) [inline torch/nn/utils/attention.py:49]: nit: for simplicity, you could combine these into one API function for subclasses to implement that is essentially `materialize_if_needed()`

File: torch/nn/utils/attention.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+    def materialize(self, device: Optional[torch.device] = None) -> torch.Tensor:
+        if device is None:
+            device = torch.device("cpu")
+        if self.variant == CausalVariant.UPPER_LEFT:
+            return self._upper_left(device)
+        elif self.variant == CausalVariant.LOWER_RIGHT:
+            return self._lower_right(device)
+
+    def needs_materialization(self) -> bool:
+        return False
```
Comment (jbschlosser) [inline torch/nn/utils/attention.py:154]: This makes it seem to me like `materialize()` never needs to be called, which can't be the case since it has an implementation. maybe all the more reason to streamline `needs_materialization()` and `materialize()` into a single API

File: torch/nn/utils/attention.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+    def dispatch(
+        query: torch.Tensor,
+        key: torch.Tensor,
+        value: torch.Tensor,
+        attn_mask,
+        dropout_p: float,
+        is_causal: bool,
+        scale: Optional[float],
+    ) -> torch.Tensor:
+        raise NotImplementedError("This is an abstract base class")
```
Comment (jbschlosser) [inline torch/nn/utils/attention.py:62]: hm it doesn't seem like this ABC is giving you much if you have to implement this function for your subclass AND manually set up a `Subclass.__torch_function__()` that calls it. It'd be nice if the base handled the `__torch_function__()` part

File: torch/nn/utils/attention.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+        query: torch.Tensor,
+        key: torch.Tensor,
+        value: torch.Tensor,
+        attn_mask: "AttnBias",
+        dropout_p: float,
+        is_causal: bool,
+        scale: Optional[float],
+    ) -> torch.Tensor:
+        raise NotImplementedError(
+            "TensorBias requires materialization, so this should never be called!"
```
Comment (jbschlosser) [inline torch/nn/utils/attention.py:88]: so is the idea that materialization and dispatching are mutually exclusive in general?

do I have it right that materialization necessarily returns a standard dense Tensor so we aren't expected to ever dispatch on some custom type?

File: test/test_transformers.py [hunk 7]
```diff
@@ -2967,6 +2971,189 @@ def rand_nt(sequence_list, num_heads, head_dim):
...
+            torch.rand, shape, device=device, dtype=torch.float16, requires_grad=True
+        )
+        query, key, value = make_tensor(), make_tensor(), make_tensor()
+        query_prototype, key_prototype, value_prototype = query_key_value_clones(query, key, value)
+        attn_bias = TensorBias(
+            torch.rand(shape.batch, shape.num_heads, shape.seq_len, shape.seq_len, dtype=torch.float16, device=device)
+        )
+
+        pytorch_output = scaled_dot_product_attention(
+            query, key, value, attn_mask=attn_bias.materialize(device), dropout_p=0.0, is_causal=False
```
Comment (jbschlosser) [inline test/test_transformers.py:3129]: This manual `materialize()` call is unnecessary right? I'd expect that the user never has to interact with materialization as a concept (unless they're implementing a custom attn bias) and that it would be tucked behind the abstraction

File: test/test_transformers.py [hunk 8]
```diff
@@ -2967,6 +2971,189 @@ def rand_nt(sequence_list, num_heads, head_dim):
...
+        torch.testing.assert_close(query.grad, query_prototype.grad, atol=grad_atol, rtol=grad_rtol)
+        torch.testing.assert_close(key.grad, key_prototype.grad, atol=grad_atol, rtol=grad_rtol)
+        torch.testing.assert_close(value.grad, value_prototype.grad, atol=grad_atol, rtol=grad_rtol)
+
+
+    @parametrize("shape", [SdpaShape(16, 16, 128, 16), SdpaShape(16, 16, 52, 32)])
+    @parametrize("compile", [True, False])
+    def test_tensor_bias(self, device, shape: SdpaShape, compile: bool):
+        if compile:
+            self.skipTest("Compiling torch_function_ not working")
```
Comment (jbschlosser) [inline test/test_transformers.py:3116]: is this still true? if so, can we make this an xfail instead so we know when to update it as this changes

**Review by jbschlosser — COMMENTED** [review]
Nice work! Got a few questions about the abstractions but it's looking good

File: test/test_transformers.py (hunk 7 above)
Comment (drisspg) [reply test/test_transformers.py:3129]: Yeah exactly, doing this here just to enforce we are testing against OG sdpa

File: torch/nn/utils/attention.py (hunk 5 above)
Comment (drisspg) [reply torch/nn/utils/attention.py:62]: ahhh good idea!

Comment (drisspg) [reply torch/nn/utils/attention.py:62]: And yeah this ABC is more of a helper "scaffolding" that I think will be helpful for people wanting to implement their own masks

File: torch/nn/utils/attention.py (hunk 4 above)
Comment (drisspg) [reply torch/nn/utils/attention.py:154]: My thinking here was, although for the standard path of SDPA you never need to materialize the bias, there may be cases where users want to interact with an instance outside of SDPA. However, at least for my implementation I made torch_function disable everything but SDPA this makes less sense..

File: torch/nn/utils/attention.py [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+
+    LOWER_RIGHT: Represents lower-right triangular bias, the include values are aligned to the lower
+    right corner of the matrix.
+    Example::
+
+        [[1, 1, 0, 0],
+         [1, 1, 1, 0],
+         [1, 1, 1, 1]]
+
+    Note that these variants are equivalent to each other when the sequence lengths of the query and key/value
```
Comment (ani300) [inline torch/nn/utils/attention.py:244]: missing part of the comment here

**Comment by drisspg** [comment]
Results from benchmarking
```Shell
+---------+--------------------+------------+-----------+-----------+-----------+-----------+----------------+----------+
|  Type   |      Speedup       | batch_size | num_heads | q_seq_len | k_seq_len | embed_dim |     dtype      | head_dim |
+---------+--------------------+------------+-----------+-----------+-----------+-----------+----------------+----------+
| Average | 1.4320973389745877 |            |           |           |           |           |                |          |
|   Max   | 2.577393980674173  |    128     |    32     |    512    |   4097    |   2048    | torch.bfloat16 |    64    |
|   Min   | 0.942474845104863  |     1      |    16     |    256    |    416    |   2048    | torch.bfloat16 |   128    |
+---------+--------------------+------------+-----------+-----------+-----------+-----------+----------------+----------+
```

File: aten/src/ATen/native/transformers/cuda/sdp_utils.cpp [hunk 10]
```diff
@@ -248,7 +248,7 @@ bool check_flash_causal_non_square_seqlens(sdp_params const& params, bool debug)
 }
 
 
-TORCH_API bool can_use_flash_attention(sdp_params const& params, bool debug) {
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.cpp:251]: rebase bug

File: aten/src/ATen/native/transformers/cuda/sdp_utils.cpp [hunk 11]
```diff
@@ -286,7 +286,7 @@ TORCH_API bool can_use_flash_attention(sdp_params const& params, bool debug) {
   }
 }
 
-TORCH_API bool can_use_mem_efficient_attention(sdp_params const& params, bool debug) {
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.cpp:289]: no it isnt'

File: aten/src/ATen/native/transformers/cuda/sdp_utils.cpp (hunk 10 above)
Comment (drisspg) [reply aten/src/ATen/native/transformers/cuda/sdp_utils.cpp:251]: no isnt

Comment (drisspg) [reply aten/src/ATen/native/transformers/cuda/sdp_utils.cpp:251]: I meant to remove these in previous layer of stack, TBH not sure how this doesnt' cause a bug will fix in this PR

File: torch/nn/utils/attention.py (hunk 2 above)
Comment (drisspg) [reply torch/nn/utils/attention.py:165]: I think your statement hit captured it perfectly, is the problem that this isn't explicitly stated ?

Comment (drisspg) [reply torch/nn/utils/attention.py:165]: In terms of overhead I meant the overhead of torch__function, which I haven't measured soo little preamture. But there is some pytree maps and I imagine for some they wouldn't wanna pay this overhead if they are using attention_masks that is just a tensor

File: torch/nn/utils/attention.py (hunk 6 above)
Comment (drisspg) [reply torch/nn/utils/attention.py:215]: I wouldn't say mutually exclusive but they are independent. the subclass defines how it should produce the sdpa_output using the `dispatch` method and it will use the base_class __torch_function__. 

There is a shortcut path for `needs_materialization` == True subclasses. I could make these classes implement dispatch. And in the dispatch make users define call `marterialize`. The idea was that users don't have to implement dispatch and just Raise the error.

File: test/test_transformers.py [hunk 12]
```diff
@@ -2967,6 +2973,134 @@ def rand_nt(sequence_list, num_heads, head_dim):
...
+    def test_tensor_bias(self, device, shape):
+        make_tensor = partial(torch.rand, shape, device=device, dtype=torch.float16, requires_grad=True)
+        attn_bias = TensorBias(
+            torch.rand(shape.batch, shape.num_heads, shape.seq_len, shape.seq_len, dtype=torch.float16, device=device)
+        )
+        self.run_test(device, False, make_tensor, make_tensor, attn_bias)
+
+    @parametrize("shape", [SdpaShape(16, 16, 128, 16), SdpaShape(16, 16, 52, 32)])
+    @xfail
+    def test_tensor_bias_compuile(self, device, shape):
```
Comment (drisspg) [inline test/test_transformers.py:3093]: spelling

File: torch/nn/utils/attention.py [hunk 13]
```diff
@@ -0,0 +1,437 @@
...
+                    is_causal=True,
+                    return_debug_mask=False,
+                    scale=og_scale,
+                )[0]
+                return _postprocess_flash_output(out, og_head_size)
+            if can_use_efficient_attention(sdpa_params):
+                compute_log_sumexp = False
+                if _input_requires_grad(query, key, value):
+                    compute_log_sumexp = True
+                return torch.ops.aten._efficient_attention_forward(
```
Comment (drisspg) [inline torch/nn/utils/attention.py:402]: @ani300 example usage

**Comment by cpuhrsch** [comment]
Small comment: I think using factory functions over `__init__` directly can help us preserve more flexibility in the future.

**Comment by drisspg** [comment]
closing in favor of https://github.com/pytorch/pytorch/pull/114823
